File: p205e/is_isomorphic.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  let stot = {};
  let ttos = {};
  for (let i = 0, len = s.length; i < len; ++i) {
    if (stot[s[i]] !== ttos[t[i]]) return false;
    stot[s[i]] = ttos[t[i]] = i;
  }
  return true;
};
// TEST
[
  ['add', 'egg', true],
  ['foo', 'bar', false],
  ['paper', 'title', true],
  ['ab', 'aa', false],
].forEach(([s, t, expected]) => {
  const actual = isIsomorphic(s, t);
  console.log(s, 'and', t, 'are isomorphic ->', actual);
  console.assert(actual === expected);
});

File: p205e/is_isomorphic.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm test
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
class Solution {
public:
    bool isIsomorphic(string s, string t)
    {
        int stMap[256] = { 0 };
        int tsMap[256] = { 0 };
        for (int i = 0; i < s.length(); ++i) {
            if (stMap[s[i]] != tsMap[t[i]])
                return false;
            stMap[s[i]] = tsMap[t[i]] = i + 1;
        }
        return true;
    }
};
struct Test {
    string s;
    string t;
    bool expected;

    void run()
    {
        Solution sol;
        bool actual = sol.isIsomorphic(s, t);
        cout << "'" << s << "' and '" << t << "' are isomorphic -> "
             << boolalpha << actual << endl;
        assert(actual == expected);
    }
};
int main(int argc, char const* argv[])
{
    vector<Test> tests = {
        { "add", "egg", true },
        { "foo", "bar", false },
        { "paper", "title", true },
        { "ab", "aa", false },
    };
    for (auto& t : tests)
        t.run();
    return 0;
}
